### DISCUSSION TITLE
Can declarative modules work with inline included files?

### DISCUSSION THREAD

**Discussion by austin362667** [PyO3/pyo3#4312] [Questions] [answered]
Consider https://pyo3.rs/v0.22.0/module#declarative-modules, for example, 

In `foo.rs`
```rust
#[pyo3::pymodule]
pub mod foo_module {
    include!("./bar.rs");
}
```

In `bar.rs`
```rust
#[pyo3::pymodule]
pub mod bar_module {
    #[pyclass]
    pub struct DataClass {
        pub hash: i64
    }
}
```

After building `foo.rs` through PyO3, can I access some structure with `#[pyclass]` in `foo_module.bar_module` in Python?

If I slighty move that structure, like following, can work.

```rust
#[pyo3::pymodule]
pub mod foo_module {
    #[pyo3::pymodule]
    pub mod bar_module {
        pub struct DataClass {
            pub hash: i64
        }
    }
}
```

**Comment by davidhewitt** [comment, marked as answer]
We can't see through the `include!` macro. 

What you can instead do is use `#[pymodule_export]`:

```rust
mod bar;

#[pyo3::pymodule]
pub mod foo_module {
    #[pymodule_export]
    use bar::bar_module;
}
```

**Reply by austin362667** [reply]
Thanks for answering so fast! It works.

Just two follow-ups.
1. I'm failing at wildcard import, like `bar::bar_module;`(imports module) or `bar::bar_module::*;`(imports every struct within the module), does it support wildcard use? If not, I want to work on a PR against this. WDYT?
2. The docs mentioned `For nested modules, the name of the parent module is automatically added.`. And I can't successfully access via the underlying complete module path while importing in Python side. Am I missing anything?

**Comment by austin362667** [comment]
Issues other people might encountered when doing something similar to my above example code snippets is well-elaborated in https://github.com/PyO3/pyo3/issues/4285. And thanks to Alex and David, it is just now fixed in https://github.com/PyO3/pyo3/pull/4288. Thanks for such a amazing works guys.
